src/otaclient_common/retry_task_map.py

```diff
@@ -195,49 +200,61 @@ def _fut_gen(self, interval: int) -> Generator[Future[Any], Any, None]:
             except Empty:
                 time.sleep(interval)
 
+    def _rtm_shutdown(self, msg: str):
+        """
+        Called by task_done_cb or dispatcher to shutdown threadpool on failure.
+        """
+        if self._rtm_lower_pool_shutdown:
+            return  # no need to shutdown again
+
+        if self._lock.acquire(blocking=False):
+            try:
+                _err_msg = f"shutdown executor and drain workitem queue: {msg}"
+                logger.warning(_err_msg)
+                self._failure_msg = _err_msg
+
+                # wait MUST be False to prevent deadlock when calling _on_shutdown from worker
+                self.shutdown(wait=False)
+                # drain the worker queues
+                with contextlib.suppress(Empty):
+                    while True:
+                        self._work_queue.get_nowait()
+            finally:
+                self._lock.release()
```

Comment:
When `_lock` is gotten by other functions like `ensure_tasks`, `self.shutdown(wait=False)` might not be called?

Reply:
Ah good catch! I should use dedicated lock for shutdown.

Reply:
Use a dedicated lock for shutdown at [f3fd928](https://github.com/tier4/ota-client/pull/504/commits/f3fd928b0a51b786b7ec5649d7645cc19e859532).

Note that `ThreadPoolExecutor` itself also has a `_shutdown_lock`, use `_rtm_shutdown_lock` to avoid conflicts.